planner/core/common_plans.go

```diff
@@ -228,6 +242,13 @@ func (e *Execute) OptimizePreparedPlan(ctx context.Context, sctx sessionctx.Cont
 				return err
 			}
 			param := prepared.Params[i].(*driver.ParamMarkerExpr)
+			if isGetVarBinaryLiteral(sctx, usingVar) {
+				binVal, convErr := val.ToBytes()
+				if convErr != nil {
+					return convErr
+				}
+				val.SetBinaryLiteral(types.BinaryLiteral(binVal))
+			}
```

Comment:
MySQL's behavior regarding supported user variables types is pretty confusing, according to its documentation 

> User variables can be assigned a value from a limited set of data types: integer, decimal, floating-point, binary or nonbinary string, or NULL value. Assignment of decimal and real values does not preserve the precision or scale of the value. A value of a type other than one of the permissible types is converted to a permissible type. For example, a value having a temporal or spatial data type is converted to a binary string. A value having the JSON data type is converted to a string with a character set of utf8mb4 and a collation of utf8mb4_bin.

> Hexadecimal or bit values assigned to user variables are treated as binary strings. To assign a hexadecimal or bit value as a number to a user variable, use it in numeric context. For example, add 0 or use CAST(... AS UNSIGNED):

We should treat it as string type actually.

Reply:
This [PR](https://github.com/pingcap/tidb/pull/4415) indicates that Hex/Bit values are represented by **TypeVarString**. In TiDB when a datum is BinaryLiteray/HexLiteral/BitLiteral, the type of corresponsding Constant expr will be TypeVarString.

So here when the variable val is setted as BinaryLiteral, it's actually a VarString.(Maybe VarString is also string type?)